bin/faucet/src/handlers.rs

```diff
@@ -55,33 +59,80 @@ pub async fn get_tokens(
 
     // Check that the amount is in the asset amount options
     if !state.config.asset_amount_options.contains(&req.asset_amount) {
-        return Err(ProcessError::BadRequest("Invalid asset amount".to_string()));
+        return Err(HandlerError::BadRequest("Invalid asset amount".to_string()));
     }
 
     let mut client = state.client.lock().await;
 
     // Receive and hex user account id
     let target_account_id = AccountId::from_hex(req.account_id.as_str())
-        .map_err(|err| ProcessError::BadRequest(err.to_string()))?;
-
-    // Execute transaction
-    info!(target: COMPONENT, "Executing mint transaction for account.");
-    let (executed_tx, created_note) = client.execute_mint_transaction(
-        target_account_id,
-        req.is_private_note,
-        req.asset_amount,
-    )?;
-
-    // Run transaction prover & send transaction to node
-    info!(target: COMPONENT, "Proving and submitting transaction.");
-    let block_height = client.prove_and_submit_transaction(executed_tx).await?;
+        .map_err(|err| HandlerError::BadRequest(err.to_string()))?;
+
+    let (created_note, block_height) = loop {
+        let mut faucet_account = client.data_store().faucet_account();
+
+        // Execute transaction
+        info!(target: COMPONENT, "Executing mint transaction for account.");
+        let (executed_tx, created_note) = client.execute_mint_transaction(
+            target_account_id,
+            req.is_private_note,
+            req.asset_amount,
+        )?;
+
+        let prev_hash = faucet_account.hash();
+
+        faucet_account
+            .apply_delta(executed_tx.account_delta())
+            .or_fail("Failed to apply faucet account delta")?;
+
+        let new_hash = faucet_account.hash();
+
+        // Run transaction prover & send transaction to node
+        info!(target: COMPONENT, "Proving and submitting transaction.");
+        match client.prove_and_submit_transaction(executed_tx.clone()).await {
+            Ok(block_height) => {
+                break (created_note, block_height);
+            },
+            Err(err) => {
```

Comment:
Is there an error variant we could match here that would specific a state mismatch? Instead of trying on all errors?

Reply:
Currently we have only "invalid argument" error for any error in this endpoint. We should improve errors returned from the block producer, but I didn't make it now in order to not conflict with your refactoring of the block producer. But we should improve it in the future, I will put `TODO` here.